crates/telio-traversal/src/session_keeper.rs

```diff
@@ -274,6 +289,25 @@ impl Runtime for State {
     where
         F: Future<Output = BoxAction<Self, std::result::Result<(), Self::Err>>> + Send,
     {
+        let mut tx_has_changed = false;
+        let mut rx_has_changed = false;
+
+        // We just care about any network activity, thus no per-peer filtering.
+        if let Some(wg) = self.network_activity_getter.as_ref() {
+            if let Ok(Some(timestamps)) = wg.get_ts().await {
+                tx_has_changed = self.last_tx_ts.map_or(false, |ts| timestamps.tx_ts > ts);
+                self.last_tx_ts = Some(timestamps.tx_ts);
+
+                rx_has_changed = self.last_rx_ts.map_or(false, |ts| timestamps.rx_ts > ts);
+                self.last_rx_ts = Some(timestamps.rx_ts);
+            }
+        }
+
+        if tx_has_changed || rx_has_changed {
```

Comment:
Hmm, shouldn't it be only `rx_has_changed` ? And is `SessionKeeper` the best entity to handle the triggering? I thought it should be also done when the connection is in `Relayed` state?

Reply:
From real life experiments both sending and receiving data causes the drainage for a while, thus we care about both of those things happening.

Now about the SessionKeeper - it **is** the best place since it was the one used for keepalives(both batched and non-batched). Batcher is only being used inside of it. For now only because we dealt with direct keepalives that were handled in SessionKeeper.

I would say that SessionKeeper outlived it's purpose and probably warrant a rename at this point though :).